From 0fbf4ce0f0cad682f3ba6e3cf1e26b3b0e9029f3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Mon, 9 Dec 2024 23:40:29 +0100
Subject: [PATCH 71/86] setupapi/tests: Make sure string is terminated.

=================================================================
==644==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffffe1fce34 at pc 0x6ffff88f2c0c bp 0x7ffffe1fb8c0 sp 0x7ffffe1fb900
READ of size 262 at 0x7ffffe1fce34 thread T0
0404:fixme:file:server_get_file_info Unsupported info class e
    #0 0x6ffff88f2c0b in strrchr /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:778:5
    #1 0x6ffff518dbfb in SetupCopyOEMInfA /home/bernhard/data/entwicklung/2024/wine/wine/dlls/setupapi\devinst.c:5552:41
    #2 0x00014004b8cb in test_copy_oem_inf /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/tests/devinst.c:4096:11
    #3 0x0001400224b5 in func_devinst /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/tests/devinst.c:4708:5
    #4 0x0001400c71ac in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #5 0x0001400c71ac in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #6 0x0001400c90df in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #7 0x6ffff80e4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #8 0x6ffff82bfa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

Address 0x7ffffe1fce34 is located in stack of thread T0 at offset 2964 in frame
    #0 0x00014004a53f in test_copy_oem_inf /home/bernhard/data/entwicklung/2024/wine\wine/dlls/setupapi/tests/devinst.c:4006

  This frame has 14 object(s):
    [32, 292) 'expect.i2138' (line 3952)
    [368, 628) 'expect.i2049' (line 3952)
    [704, 1228) 'orig_info.i' (line 3961)
    [1360, 1364) 'size.i' (line 3963)
    [1376, 1636) 'expect.i1501' (line 3952)
    [1712, 1972) 'expect.i' (line 3952)
    [2048, 2568) 'path' (line 4007)
    [2704, 2964) 'dest' (line 4007)
    [3040, 3300) 'orig_dest' (line 4007) <== Memory access at offset 2964 partially underflows this variable
    [3376, 3636) 'orig_cwd' (line 4008)
    [3712, 3720) 'filepart' (line 4008)
    [3744, 4004) 'pnf' (line 4008)
    [4080, 4128) 'system_info' (line 4009)
    [4160, 4164) 'size' (line 4011)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp, SEH and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/bernhard/data/entwicklung/2024/wine/wine/dlls/setupapi\devinst.c:5552:41 in SetupCopyOEMInfA
Shadow bytes around the buggy address:
  0x7ffffe1fcb80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1fcc00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1fcc80: 00 00 00 00 00 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2
  0x7ffffe1fcd00: f2 f2 f2 f2 f2 f2 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1fcd80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x7ffffe1fce00: 00 00 00 00 00 00[04]f2 f2 f2 f2 f2 f2 f2 f2 f2
  0x7ffffe1fce80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1fcf00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1fcf80: 04 f2 f2 f2 f2 f2 f2 f2 f2 f2 00 00 00 00 00 00
  0x7ffffe1fd000: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1fd080: 00 00 00 00 00 00 00 00 00 00 04 f2 f2 f2 f2 f2
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==644==ABORTING
make: *** [Makefile:352637: dlls/setupapi/tests/x86_64-windows/devinst.ok] Fehler 1
0404:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 00007FFFFEA8FE80
---
 dlls/setupapi/tests/devinst.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/dlls/setupapi/tests/devinst.c b/dlls/setupapi/tests/devinst.c
index 3b900eeed31..1a369f9c490 100644
--- a/dlls/setupapi/tests/devinst.c
+++ b/dlls/setupapi/tests/devinst.c
@@ -4092,8 +4092,9 @@ static void test_copy_oem_inf(struct testsign_context *ctx)
 
     /* Test with a relative path. */
     SetLastError(0xdeadbeef);
-    memset(dest, 0xcc, sizeof(dest));
-    ret = SetupCopyOEMInfA("winetest.inf", NULL, 0, SP_COPY_NOOVERWRITE, dest, sizeof(dest), NULL, &filepart);
+    memset(dest, 0xcc, sizeof(dest) - 1);
+    dest[sizeof(dest) - 1] = '\0';
+    ret = SetupCopyOEMInfA("winetest.inf", NULL, 0, SP_COPY_NOOVERWRITE, dest, sizeof(dest) - 1, NULL, &filepart);
     ok(ret == TRUE, "Got %d.\n", ret);
     ok(!GetLastError(), "Got error %#lx.\n", GetLastError());
     ok(file_exists("winetest.inf"), "Expected source inf to exist.\n");
-- 
2.47.1

